### PR TITLE
Remove dumbo's dependency on mmds

### PR DESCRIPTION
**Pull request by moricho**
Signed-off-by: moricho <ikeda.morito@gmail.com>

## Reason for This PR

Fix: https://github.com/firecracker-microvm/firecracker/issues/1770

## Description of Changes

This PR 
・moves `dumbo::ns` to `mmds` crate.
・implements response retrieval callback passed to dumbo's tcp::handler
to remove the dumbo's dependency on `mmds`.

- [ ] This functionality can be added in [`rust-vmm`](https://github.com/rust-vmm).

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license.

## PR Checklist

`[Author TODO: Meet these criteria.]`
`[Reviewer TODO: Verify that these criteria are met. Request changes if not]`

- [ ] All commits in this PR are signed (`git commit -s`).
- [ ] The reason for this PR is clearly provided (issue no. or explanation).
- [ ] The description of changes is clear and encompassing.
- [ ] Any required documentation changes (code and docs) are included in this PR.
- [ ] Any newly added `unsafe` code is properly documented.
- [ ] Any API changes are reflected in `firecracker/swagger.yaml`.
- [ ] Any user-facing changes are mentioned in `CHANGELOG.md`.

### CODE REVIEW

File: src/dumbo/src/tcp/handler.rs [hunk 1]
```diff
@@ -148,6 +149,48 @@ enum RecvSegmentOutcome {
     UnexpectedSegment(bool),
 }
 
+/// build_response build micro_http::Response with given http version, status code, and body.
+///
```
Comment (iulianbarbu) [inline src/dumbo/src/tcp/handler.rs:153]: Remove this empty comment line.

File: src/dumbo/src/tcp/handler.rs [hunk 2]
```diff
@@ -148,6 +149,48 @@ enum RecvSegmentOutcome {
     UnexpectedSegment(bool),
 }
 
+/// build_response build micro_http::Response with given http version, status code, and body.
```
Comment (iulianbarbu) [inline src/dumbo/src/tcp/handler.rs:152]:
```suggestion
/// `build_response` builds the `micro_http::Response` with a given HTTP version, status code, and body.
```

File: src/dumbo/src/tcp/handler.rs [hunk 3]
```diff
@@ -148,6 +149,48 @@ enum RecvSegmentOutcome {
...
+///
+pub fn build_response(http_version: Version, status_code: StatusCode, body: Body) -> Response {
+    let mut response = Response::new(http_version, status_code);
+    response.set_body(body);
+    response
+}
+
+/// parse_request parses the request and build response by given callback function.
+///
```
Comment (iulianbarbu) [inline src/dumbo/src/tcp/handler.rs:161]: Remove this empty comment line.

File: src/dumbo/src/tcp/handler.rs [hunk 4]
```diff
@@ -148,6 +149,48 @@ enum RecvSegmentOutcome {
...
 
+/// build_response build micro_http::Response with given http version, status code, and body.
+///
+pub fn build_response(http_version: Version, status_code: StatusCode, body: Body) -> Response {
+    let mut response = Response::new(http_version, status_code);
+    response.set_body(body);
+    response
+}
+
+/// parse_request parses the request and build response by given callback function.
```
Comment (iulianbarbu) [inline src/dumbo/src/tcp/handler.rs:160]:
```suggestion
/// `parse_request` parses the request bytes and builds a `micro_http::Response` by the given callback function.
```

File: src/dumbo/src/tcp/handler.rs [hunk 5]
```diff
@@ -148,6 +149,48 @@ enum RecvSegmentOutcome {
...
+                Version::default(),
+                StatusCode::BadRequest,
+                Body::new("Invalid request.".to_string()),
+            ),
+            RequestError::InvalidHeader => build_response(
+                Version::default(),
+                StatusCode::BadRequest,
+                Body::new("Invalid headers.".to_string()),
+            ),
+            RequestError::UnsupportedHeader => unreachable!(),
```
Comment (iulianbarbu) [inline src/dumbo/src/tcp/handler.rs:189]: This will panic if the `Request` will contain an unsupported header. This can happen when trying to parse a `Request` and the `Content-type` is set to a content type different than `text/plain` or `application/json`. Can we return here a `BadRequest` error response as well?

File: src/dumbo/src/tcp/handler.rs [hunk 6]
```diff
@@ -488,10 +532,87 @@ impl TcpIPv4Handler {
...
+        let response_body = String::from("test");
+        build_response(
+            request.http_version(),
+            StatusCode::OK,
+            Body::new(response_body),
+        )
+    }
+
+    #[test]
+    fn test_parse_request() {
```
Comment (iulianbarbu) [inline src/dumbo/src/tcp/handler.rs:559]: Please add a test for `UnsupportedHeader` case as well .

**Comment by moricho** [comment]
@iulianbarbu Thanks for your review! I fixed. In addition, I moved `persist` module from `dumbo` to `mmds` which was implemented in https://github.com/firecracker-microvm/firecracker/pull/1773 

ref: https://github.com/firecracker-microvm/firecracker/pull/1813/commits/7de102fbf3fe63c8afa7c29ff956939fe291e686

File: src/dumbo/src/tcp/handler.rs [hunk 7]
```diff
@@ -148,6 +149,46 @@ enum RecvSegmentOutcome {
     UnexpectedSegment(bool),
 }
 
+/// `build_response` builds the `micro_http::Response` with a given HTTP version, status code, and body.
```
Comment (iulianbarbu) [inline src/dumbo/src/tcp/handler.rs:152]: Why is this needed?

File: src/mmds/src/lib.rs [hunk 8]
```diff
@@ -5,15 +5,24 @@
...
 
 pub mod data_store;
+pub mod ns;
+pub mod persist;
 
 use serde_json::{Map, Value};
 use std::sync::{Arc, Mutex};
 
 use data_store::{Error as MmdsError, Mmds};
-use micro_http::{Body, Request, RequestError, Response, StatusCode, Version};
```
Comment (iulianbarbu) [inline src/mmds/src/lib.rs:16]: My thought is that `build_response` function need to be implemented at the level of `mmds` crate. As you've provided right now, we would need to consume `build_response` from `dumbo` to create `micro-http::Response`s, which is not ergonomic when `dumbo::build_response` would need to be extended, to add other headers lets say, to the provided response. For example, if one needs to add an extra `micro-http::Header` to the response, it will need to call first `dumbo::build_response`, then take a mutable reference to it, and then add the header. Another option will be to adjust this `dumbo::build_response` function for multiple response building cases but I do not think it is in the scope of the `dumbo` crate. `micro-http::Response` already provides this API for building responses for multiple cases, so reinventing a build response API in `dumbo` does not worth it.  I suggest to make this response builder function implementation specific. What do you think?

**Comment by iulianbarbu** [comment]
Hey, @moricho ! Let us know if you need any help with this.

**Comment by moricho** [comment]
@iulianbarbu I’m sorry for the late reply! I'll start to work on this issue again soon :) Just a moment, please.

**Comment by iulianbarbu** [comment]
> @iulianbarbu I’m sorry for the late reply! I'll start to work on this issue again soon :) Just a moment, please. Thank you for your help.

Cool. No problem. Take your time. Let us know if we can help. :)

File: src/mmds/src/lib.rs (hunk 8 above)
Comment (moricho) [reply src/mmds/src/lib.rs:16]: @iulianbarbu 
I agree with your thought. Indeed I implemented `build_response` function in `dumbo` crate only for using it in `dumbo::parse_request` function. If we implement `build_response` in `mmds` crate, how should we pass `mmds::build_response` to `dumbo::parse_request`?
I think the first way to do that is passing `mmds::build_response` through one of the argument like `callback` as I've already implemented. Do you have any other ideas?

**Comment by moricho** [comment]
@iulianbarbu Thank you for waiting! I resolved conflicts for now and replied to your comment(https://github.com/firecracker-microvm/firecracker/pull/1813#discussion_r417984014)

File: src/mmds/src/lib.rs (hunk 8 above)
Comment (iulianbarbu) [reply src/mmds/src/lib.rs]: Passing the response through the callback is perfectly fine, as far as I can tell. I do not have any issue with it.

File: src/dumbo/src/tcp/handler.rs [hunk 9]
```diff
@@ -180,12 +181,33 @@ impl TcpIPv4Handler {
         self.local_ipv4_addr = ipv4_addr;
     }
 
+    /// Getter for the local IPv4 address of this TCP handler.
+    pub fn get_local_ipv4_addr(&self) -> Ipv4Addr {
+        self.local_ipv4_addr
+    }
+
+    /// Getter for the local port of this TCP handler.
```
Comment (lauralt) [inline src/dumbo/src/tcp/handler.rs:189]: There is no need to include `getter` in the comment. This comment applies for the other getters as well.

```suggestion
    /// Returns the local port of this TCP handler.
```

File: src/dumbo/src/tcp/handler.rs [hunk 10]
```diff
@@ -493,6 +516,16 @@ mod tests {
         }
     }
 
+    fn mock_callback(request: Result<Request, RequestError>) -> Response {
```
Comment (lauralt) [inline src/dumbo/src/tcp/handler.rs:519]: If we want to create a mock callback, wouldn't it be enough to return some `Response` type (for example `Response::new(Version::Http11, StatusCode::OK)`) and not even check the `request` variant via that match?

Also, this function looks like a duplicate. Can we get rid of one of them?

File: src/mmds/src/lib.rs [hunk 11]
```diff
@@ -31,6 +39,13 @@ impl Into<OutputFormat> for MediaType {
     }
 }
 
+/// `build_response` builds the `micro_http::Response` with a given HTTP version, status code, and body.
+pub fn build_response(http_version: Version, status_code: StatusCode, body: Body) -> Response {
```
Comment (lauralt) [inline src/mmds/src/lib.rs:43]: Does this need to be `pub`?

File: src/dumbo/src/tcp/endpoint.rs [hunk 12]
```diff
@@ -299,16 +303,36 @@ impl Endpoint {
     }
 }
 
+/// `parse_request` parses the request bytes and builds a `micro_http::Response` by the given callback function.
+pub fn parse_request(
+    request_bytes: &[u8],
+    callback: fn(Result<Request, RequestError>) -> Response,
+) -> Response {
```
Comment (iulianbarbu) [inline src/dumbo/src/tcp/endpoint.rs:310]: This `parse_request` needs to map a stream of bytes to a `micro_http::Request`. If any error is raised during this mapping, this error should be sent to upper levels. 

```suggestion
) -> Result<Response, RequestError> {
```

File: src/dumbo/src/tcp/endpoint.rs [hunk 13]
```diff
@@ -299,16 +303,36 @@ impl Endpoint {
...
+    let request = Request::try_from(request_bytes);
```
Comment (iulianbarbu) [inline src/dumbo/src/tcp/endpoint.rs:311]:
```suggestion
    let request = Request::try_from(request_bytes)?;
```

File: src/dumbo/src/tcp/endpoint.rs [hunk 14]
```diff
@@ -299,16 +303,36 @@ impl Endpoint {
...
+    callback(request)
```
Comment (iulianbarbu) [inline src/dumbo/src/tcp/endpoint.rs:312]:
```suggestion
    Ok(callback(request))
```

File: src/dumbo/src/tcp/endpoint.rs [hunk 15]
```diff
@@ -299,16 +303,36 @@ impl Endpoint {
...
     use super::*;
 
     use std::fmt;
     use std::str::from_utf8;
 
+    use micro_http::{Request, RequestError, Response, StatusCode, Version};
     use pdu::tcp::Flags as TcpFlags;
     use tcp::connection::tests::ConnectionTester;
 
+    fn mock_callback(request: Result<Request, RequestError>) -> Response {
```
Comment (iulianbarbu) [inline src/dumbo/src/tcp/endpoint.rs:326]: Even the mock callback should process an already obtained `micro_http::Request` and return a correspondent response.
```suggestion
    fn mock_callback(request: Request) -> Response {
```

File: src/dumbo/src/tcp/endpoint.rs [hunk 16]
```diff
@@ -299,16 +303,36 @@ impl Endpoint {
...
+        match request {
+            Ok(request) => {
+                let _uri = request.uri().get_abs_path();
+                Response::new(Version::Http11, StatusCode::OK)
+            }
+            Err(_) => Response::new(Version::Http11, StatusCode::BadRequest),
+        }
```
Comment (iulianbarbu) [inline src/dumbo/src/tcp/endpoint.rs:333]:
```suggestion
            // If we really need some request based processing in the mock callback, we should do it here.
           // Otherwise, return a default response.
           Response::new(Version::Http11, StatusCode::OK)
        }
```

**Comment by dianpopa** [comment]
If I understand correctly issue #1770 aims at decoupling the specific mmds logic that is found inside the dumbo crate. Another module that is totally specific to Firecracker's mmds is also the `endpoint` module. See this [doc comment](https://github.com/firecracker-microvm/firecracker/pull/1813/files#diff-3b36cb6cd2f1ea7852b501d50ad0ad32R4). So putting a callback in the receive_packet function does not solve our desire to make dumbo Firecracker independent. 
What I see as a clean fix for this is to have an endpoint interface defined in dumbo that asks the implementing crate for the needed functionality. Another thing that is Firecracker specific is `struct ConnectionTuple` which acts as a way of uniquely identifying an `Endpoint`:
So what I suggest is this:
* the handler module offers a `Endpoint` interface:
```
pub trait Endpoint {
    fn get_id(&self) -> EndpointId;
    fn receive_segment<T: NetworkBytes>(
        &mut self,
        s: &TcpSegment<T>) -> Response;
    ...
}
```
`EndpointId `will play the role of `ConnectionTuple` and it could be
`pub type EndpointId = u64;` the hash of the unique string that would result by appending the remote address and port (the old fields of ConnectionTuple).
* endpoint module from dumbo moves to mmds as an implementation for the endpoint interface from handler.rs
* TcpIPv4Handler becomes this:

```
pub struct TcpIPv4Handler {
    // Handler IPv4 address used for every connection.
    local_ipv4_addr: Ipv4Addr,
    // Handler TCP port used for every connection.
    local_port: u16,
    // This map holds the currently active endpoints, identified by their connection tuple.
    endpoints: HashMap<EndpointId, dyn Endpoint>,
    // Maximum number of concurrent connections we are willing to handle.
    max_connections: usize,
    // Holds connections which are able to send segments immediately.
    active_connections: HashSet<EndpointId>,
    // Remembers the closest timestamp into the future when one of the connections has to deal
    // with an RTO trigger.
    next_timeout: Option<(u64, EndpointId)>,
    // RST segments awaiting to be sent.
    rst_queue: Vec<(EndpointId, RstConfig)>,
    // Maximum size of the RST queue.
    max_pending_resets: usize,
}
```
WDYT? is this worth trying out?

**Comment by iulianbarbu** [comment]
> If I understand correctly issue #1770 aims at decoupling the specific mmds logic that is found inside the dumbo crate. Another module that is totally specific to Firecracker's mmds is also the `endpoint` module. See this [doc comment](https://github.com/firecracker-microvm/firecracker/pull/1813/files#diff-3b36cb6cd2f1ea7852b501d50ad0ad32R4). So putting a callback in the receive_packet function does not solve our desire to make dumbo Firecracker independent.
> What I see as a clean fix for this is to have an endpoint interface defined in dumbo that asks the implementing crate for the needed functionality. Another thing that is Firecracker specific is `struct ConnectionTuple` which acts as a way of uniquely identifying an `Endpoint`:
> So what I suggest is this:
> 
>     * the handler module offers a `Endpoint` interface:
> 
> 
> ```
> pub trait Endpoint {
>     fn get_id(&self) -> EndpointId;
>     fn receive_segment<T: NetworkBytes>(
>         &mut self,
>         s: &TcpSegment<T>) -> Response;
>     ...
> }
> ```
> 
> `EndpointId `will play the role of `ConnectionTuple` and it could be
> `pub type EndpointId = u64;` the hash of the unique string that would result by appending the remote address and port (the old fields of ConnectionTuple).
> 
>     * endpoint module from dumbo moves to mmds as an implementation for the endpoint interface from handler.rs
> 
>     * TcpIPv4Handler becomes this:
> 
> 
> ```
> pub struct TcpIPv4Handler {
>     // Handler IPv4 address used for every connection.
>     local_ipv4_addr: Ipv4Addr,
>     // Handler TCP port used for every connection.
>     local_port: u16,
>     // This map holds the currently active endpoints, identified by their connection tuple.
>     endpoints: HashMap<EndpointId, dyn Endpoint>,
>     // Maximum number of concurrent connections we are willing to handle.
>     max_connections: usize,
>     // Holds connections which are able to send segments immediately.
>     active_connections: HashSet<EndpointId>,
>     // Remembers the closest timestamp into the future when one of the connections has to deal
>     // with an RTO trigger.
>     next_timeout: Option<(u64, EndpointId)>,
>     // RST segments awaiting to be sent.
>     rst_queue: Vec<(EndpointId, RstConfig)>,
>     // Maximum size of the RST queue.
>     max_pending_resets: usize,
> }
> ```
> 
> WDYT? is this worth trying out?

Not sure if this is worth trying out for this specific issue. As far as #1770 goes, the main topic, which I've thought about, was to remove the unnecessary dumbo dependency on MMDS.

One the other hand, while looking into your investigation, it seems like an interesting direction for further thought and maybe prototype. The dumbo's endpoint dependency on MMDS logic can be seen rather as a theoretical dependency, because of the fact that dumbo is designed to provide the necessary network stack for Ethernet/IP/TCP/HTTP communication. This shouldn't be necessary the case, because dumbo's Ethernet/IP/TCP dumbo network stack can be used also with an arbitrary application protocol (not necessary HTTP, required for MMDS). Hence, you propose an interface to be implemented by applications, for their network stack specific use case, where they can assert on their preference for the application protocol. This can be also generalized for the network protocol or for the link-layer protocol, but I can not tell if there is any data for such a need, rather than generalizing/decoupling dumbo's implementation and making it attractive for similar Firecracker use cases.

Please, let me know your thoughts.

File: src/mmds/src/lib.rs [hunk 17]
```diff
@@ -59,14 +74,7 @@ pub fn json_patch(target: &mut Value, patch: &Value) {
...
 
-fn build_response(http_version: Version, status_code: StatusCode, body: Body) -> Response {
-    let mut response = Response::new(http_version, status_code);
-    response.set_body(body);
-    response
-}
-
-pub fn parse_request(request_bytes: &[u8]) -> Response {
-    let request = Request::try_from(request_bytes);
+pub fn callback(request: Result<Request, RequestError>) -> Response {
```
[the remark this answers is not in the document]
Comment (lauralt) [reply src/mmds/src/lib.rs:77]: +1 on renaming `callback()`. From what I see, this function is used in `parse_request()`, maybe it would make sense to rename that function to `parse_request_bytes()` and this one to `convert_to_response()`.

File: src/dumbo/src/tcp/endpoint.rs (hunk 13 above)
Comment (lauralt) [reply src/dumbo/src/tcp/endpoint.rs:311]: Hmm, it seems that this way we won't build the responses that were built before in the `Err(e)` match branch from `parse_request()` (or `callback()` now).
Are you thinking we should move that part somewhere else? It seems okay to build all the possible responses in the same place; maybe `callback()` should be somehow split in smaller functions, but I don't think we should treat those "error" cases in an upper level (as you mentioned I guess in a comment above).
Sorry in advance if I'm missing something, it's kind of the first time I'm looking at this code :sweat:.

File: src/mmds/src/lib.rs [hunk 18]
```diff
@@ -156,19 +164,28 @@ pub fn parse_request(request_bytes: &[u8]) -> Response {
...
+        let actual_response = parse_request(request.as_bytes(), callback);
         assert!(expected_response.status() == actual_response.status());
         assert!(expected_response.body().unwrap() == actual_response.body().unwrap());
     }
 
+    fn check_callback_response(actual: Response, expected: Response) {
+        assert!(expected.status() == actual.status());
+        assert!(expected.body().unwrap() == actual.body().unwrap());
+        assert!(expected.http_version() == actual.http_version());
+    }
```
Comment (lauralt) [inline src/mmds/src/lib.rs:185]: This looks like a `PartialEq` implementation. Would it make sense to `impl PartialEq` here instead?

**Comment by iulianbarbu** [comment]
Hey, @moricho ! How it's going? Are you planning to address the feedback anytime soon? We would be glad to take a stab on it, to address what is left, if you are out of bandwidth. Let us know if that's OK.

**Comment by moricho** [comment]
@iulianbarbu Thanks for reminding me. I'll come back and address the feedback soon!

File: src/dumbo/src/tcp/endpoint.rs (hunk 13 above)
Comment (lauralt) [reply src/dumbo/src/tcp/endpoint.rs:311]: Sorry for the late reply, your proposal sounds good! From what it seems, there shouldn't be a lot of effort involved, at least from the implementation point of view; in tests, on the other hand, I think there will be several changes involved :-?. @moricho what do you think about this proposal? Does it make sense to you?

Comment (moricho) [reply src/dumbo/src/tcp/endpoint.rs:311]: @iulianbarbu @lauralt It makes sense to me. Thanks for the detailed explanation! Based on your proposal, I'm now refactoring that part. 
In that regard, could I ask you to give me some advice?  Do you think it's reasonable that we implement a function to build error response at `endpoint` level to treat `parse_request` error, aside from existing one to build mmds response (`build_response` func) at `mmds` crate ?

Comment (iulianbarbu) [reply src/dumbo/src/tcp/endpoint.rs:311]: @moricho , build response at the endpoint level makes sense for me, for responding to HTTP request parse errors.

I see it like this:

Dumbo assures the HTTP request is valid and if so it is passing it to MMDS. If the HTTP request is not valid, dumbo will emit a response corresponding to the `RequestError` encountered while parsing.

Before your PR, `RequestError` handling was done inside MMDS, where the request buffer was also parsed. Then, `mmds::build_response` was returning MMDS related errors (e.g `MmdsError::NotFound`) and micro-http request parsing error as well (e.g. `RequestError::InvalidUri`). Now, because we want to move the HTTP requests parsing in dumbo, we have to move `RequestError` handling (building and sending an HTTP response in case of parse error) in dumbo. Here, we can use a `build_response` function similar to `build_response` function from MMDS.

Let me know if this makes sense.

Comment (moricho) [reply src/dumbo/src/tcp/endpoint.rs:311]: @iulianbarbu Thanks for your advice! It makes sense for me. Based on this discussion, I'll do refactoring continuously.

File: src/dumbo/src/tcp/endpoint.rs [hunk 19]
```diff
@@ -522,4 +586,72 @@ mod tests {
...
+        // Test invalid HTTP format.
+        let request_bytes = b"GET / HTTP/1.1\r\n";
+        let mut expected_response = Response::new(Version::Http11, StatusCode::BadRequest);
+        expected_response.set_body(Body::new("Invalid request.".to_string()));
+        let actual_response = parse_request_bytes(request_bytes, mock_callback);
+        assert!(actual_response == expected_response);
+    }
+
+    #[test]
+    fn test_unsupported_http_method() {
```
Comment (iulianbarbu) [inline src/dumbo/src/tcp/endpoint.rs:615]: This test is not relevant for Dumbo, but it is relevant for MMDS. Can you move it back to MMDS, please? Have in mind also that this test, in MMDS, was checking against `PUT` and `PATCH`, which were not allowed methods, in that context.

Comment (moricho) [reply src/dumbo/src/tcp/endpoint.rs]: Thank you for comment. I have moved `test_unsupported_http_method ` from `dumbo` to `mmds` and also added `PUT` and `PATCH` cases.

File: src/mmds/src/lib.rs (hunk 18 above)
Comment (moricho) [reply src/mmds/src/lib.rs]: @lauralt Thank you for review! That makes sense. I added `#[derive(PartialEq)]` for the necessary part and remove `check_callback_response` function.

File: src/mmds/src/lib.rs [hunk 20]
```diff
@@ -78,108 +87,80 @@ fn sanitize_uri(mut uri: String) -> String {
...
-        let request = format!("{} http://169.254.169.254/ HTTP/1.1\r\n\r\n", method);
+        let request_string = format!("{} http://169.254.169.254/ HTTP/1.1\r\n\r\n", method);
+        let request_bytes = request_string.as_bytes();
         let mut expected_response = Response::new(Version::Http11, status);
         expected_response.set_body(Body::new(err_msg));
-        let actual_response = parse_request(request.as_bytes());
+        if status == StatusCode::MethodNotAllowed {
+            expected_response.allow_method(Method::Get);
+        }
+        expected_response.allow_method(Method::Get);
```
Comment (iulianbarbu) [inline src/mmds/src/lib.rs:162]: These lines conflicts themselves. You add an `Allow` header even if the expected response should not contain it (e.g. check method failure for "GET", StatusCode::Ok, "dont care"). Furthermore, the added Allow header is not asserted against the actual response.

```suggestion
        if status == StatusCode::MethodNotAllowed {
            expected_response.allow_method(Method::Get);
        }
```

File: src/mmds/src/lib.rs [hunk 21]
```diff
@@ -78,108 +87,80 @@ fn sanitize_uri(mut uri: String) -> String {
...
+        let actual_response = parse_request_bytes(request_bytes, convert_to_response);
```
Comment (iulianbarbu) [inline src/mmds/src/lib.rs:163]:
```suggestion
        let actual_response = parse_request_bytes(request_bytes, convert_to_response);
        assert!(expected_response.allow() == actual_response.allow())
```

**Comment by iulianbarbu** [comment]
For a better history, I would also squash all the commits inside the first one (`dumbo: remove dumbo's dependency on mmds`).

File: src/mmds/src/lib.rs [hunk 22]
```diff
@@ -78,110 +87,80 @@ fn sanitize_uri(mut uri: String) -> String {
...
-                    Body::new("Not allowed HTTP method."),
-                );
-                response.allow_method(Method::Get);
-                return response;
-            }
-
-            // The data store expects a strict json path, so we need to
-            // sanitize the URI.
-            let json_pointer = sanitize_uri(uri.to_string());
+pub fn convert_to_response(request: Request) -> Response {
```
Comment (lauralt) [inline src/mmds/src/lib.rs:90]: This doesn't need to be `pub`.

File: src/mmds/src/lib.rs [hunk 23]
```diff
@@ -78,110 +87,80 @@ fn sanitize_uri(mut uri: String) -> String {
...
     }
 }
 
 #[cfg(test)]
 mod tests {
     extern crate serde_json;
+
     use super::*;
+    use dumbo::tcp::parse_request_bytes;
+    use micro_http::{Request, Response, StatusCode, Version};
```
Comment (lauralt) [inline src/mmds/src/lib.rs:152]:
```suggestion
    use micro_http::Version;
```

File: src/mmds/src/lib.rs [hunk 24]
```diff
@@ -78,110 +87,80 @@ fn sanitize_uri(mut uri: String) -> String {
...
         let mut expected_response = Response::new(Version::Http11, status);
         expected_response.set_body(Body::new(err_msg));
-        let actual_response = parse_request(request.as_bytes());
-        assert!(expected_response.status() == actual_response.status());
-        assert!(expected_response.body().unwrap() == actual_response.body().unwrap());
+        if status == StatusCode::MethodNotAllowed {
+            expected_response.allow_method(Method::Get);
+        }
+        let actual_response = parse_request_bytes(request_bytes, convert_to_response);
+        assert!(expected_response.allow() == actual_response.allow());
```
Comment (lauralt) [inline src/mmds/src/lib.rs:163]:
```suggestion
        assert_eq!(expected_response.allow(), actual_response.allow());
```

File: src/mmds/src/lib.rs [hunk 25]
```diff
@@ -224,54 +203,44 @@ mod tests {
...
-        let request = b"HTTP/1.1";
-        let dummy_response = Response::new(Version::Http11, StatusCode::BadRequest);
-        assert!(parse_request(request).status() == dummy_response.status());
+        // Test resource not found.
+        let request_bytes = b"GET http://169.254.169.254/invalid HTTP/1.0\r\n\r\n";
+        let request = Request::try_from(request_bytes).unwrap();
+        let mut expected_response = Response::new(Version::Http10, StatusCode::NotFound);
+        expected_response.set_body(Body::new("Resource not found: /invalid.".to_string()));
+        let actual_response = convert_to_response(request);
+        assert!(actual_response == expected_response);
```
Comment (lauralt) [inline src/mmds/src/lib.rs:212]: nit:
```suggestion
        assert_eq!(actual_response, expected_response);
```
Applies to the entire test.

File: src/mmds/src/lib.rs [hunk 26]
```diff
@@ -224,54 +203,44 @@ mod tests {
...
 
-        // Test unsupported HTTP version.
-        let request = b"GET http://169.254.169.255/ HTTP/2.0\r\n\r\n";
+        // Test NotImplemented
```
Comment (lauralt) [inline src/mmds/src/lib.rs:214]: nit:
```suggestion
        // Test NotImplemented.
```

File: src/dumbo/src/tcp/endpoint.rs [hunk 27]
```diff
@@ -299,15 +304,62 @@ impl Endpoint {
     }
 }
 
+fn build_response(http_version: Version, status_code: StatusCode, body: Body) -> Response {
```
[the remark this answers is not in the document]
Comment (iulianbarbu) [reply src/dumbo/src/tcp/endpoint.rs:307]: I can not find where `parse_request_bytes()` is used outside `dumbo` crate. Can you please tell where is the usage?

There are two drawbacks that I see for keeping one single `build_response`:
1) `build_response` from `dumbo` can change/be deleted and MMDS is impacted.
2) `build_response` is a wrapper over `micro_http::Request` building. Using it inside MMDS requires to make it a public method. This would blur the dumbos API by exposing a non-`dumbo` functionality.

Although they are duplicate, I would argue that keeping them duplicate enforces better crates separation, which is preferred. Please take a look at #1819 , which is introduced by a similar scenario. There, MMDS had a dependency on micro-http inner implementation, which assured that every HTTP request is a GET. However, this implementation had changed and the previous assumption did not hold anymore, resulting in that issue.

File: src/dumbo/src/tcp/mod.rs [hunk 28]
```diff
@@ -92,6 +93,12 @@ pub fn seq_at_or_after(a: Wrapping<u32>, b: Wrapping<u32>) -> bool {
 #[cfg(test)]
 mod tests {
     use super::*;
+    use micro_http::{Request, Response, StatusCode, Version};
+
+    pub fn mock_callback(request: Request) -> Response {
```
[the remark this answers is not in the document]
Comment (iulianbarbu) [reply src/dumbo/src/tcp/mod.rs:98]: My 2 cents: The comment says the same thing as the code :)

pub - expect usage outside the module.
mock_.. - compile the minimum functionality to produce the needed side effect, even if it is dummy.

File: src/dumbo/src/tcp/endpoint.rs (hunk 27 above)
[the remark this answers is not in the document]
Comment (lauralt) [reply src/dumbo/src/tcp/endpoint.rs:307]: > I can not find where `parse_request_bytes()` is used outside `dumbo` crate. Can you please tell where is the usage?

https://github.com/moricho/firecracker/commit/68b6c0ab22cfa69b20ba39e1156d2f1c668b1a2b#diff-7d882cb8cb3286a3411207ff13d0c72bR162

and the import is here: https://github.com/moricho/firecracker/commit/68b6c0ab22cfa69b20ba39e1156d2f1c668b1a2b#diff-f7e4add39fc8dbdc081b77a61f5c586bR10 :(

> There are two drawbacks that I see for keeping one single `build_response`:
> 
> 1. `build_response` from `dumbo` can change/be deleted and MMDS is impacted.
> 2. `build_response` is a wrapper over `micro_http::Request` building. Using it inside MMDS requires to make it a public method. This would blur the dumbos API by exposing a non-`dumbo` functionality.
> 
> Although they are duplicate, I would argue that keeping them duplicate enforces better crates separation, which is preferred. Please take a look at #1819 , which is introduced by a similar scenario. There, MMDS had a dependency on micro-http inner implementation, which assured that every HTTP request is a GET. However, this implementation had changed and the previous assumption did not hold anymore, resulting in that issue.

Hmm, the situations are somehow not the same; assuring that every HTTP request is a GET sounds like something MMDS-specific, therefore the check/enforcement should be done at MMDS level, here `build_response` doesn't contain any dumbo/MMDS logic, it's more of a helper function, so that's why I suggested to leave just one. The only thing that I would expect to be different between a response that is built at MMDS level and one that is built at dumbo level is related to the allowed method(s). I agree it's weird to expose a non-dumbo function via the dumbo API, I don't like that either, we should definitely avoid it, another suggestion that I had in mind when I left the comment was to move it to `micro_http` since it's a pretty basic/generic function that can build a response based on some input.
We can also keep them duplicate to be super safe, but I would've liked to avoid that.

Also, it would be great if we could avoid using `parse_request_bytes` in MMDS tests. Btw, just curious, `MethodNotAllowed` seems already tested in `test_convert_to_response()`, why do we test it again in `test_unsupported_http_method`? Also, it looks like we are testing in MMDS the `micro_http` unsupported methods. It makes sense to test them again here as a sanity check since we must make sure that only `GET` is supported, but the problem that I see is that we don't actually check in MMDS for the `NotImplemented` methods, we kind of rely on dumbo to do the proper checks and build the right response. I would drop `test_unsupported_http_method()` until we make a clear separation here between dumbo and MMDS.

File: src/dumbo/src/tcp/mod.rs (hunk 28 above)
[the remark this answers is not in the document]
Comment (lauralt) [reply src/dumbo/src/tcp/mod.rs:98]: I'm fine either way :+1:. This function looks a bit random here and wanted to somehow explain what's up with it.

**Comment by moricho** [comment]
@lauralt Thank you for review. I fixed some parts you mentioned.

@iulianbarbu @lauralt 
I'll reply to those comment and subsequent discussion later :)
・https://github.com/firecracker-microvm/firecracker/pull/1813#discussion_r448355890
・https://github.com/firecracker-microvm/firecracker/pull/1813#discussion_r448406433

File: src/dumbo/src/tcp/endpoint.rs (hunk 27 above)
[the remark this answers is not in the document]
Comment (lauralt) [reply src/dumbo/src/tcp/endpoint.rs:307]: And regarding `build_response()`, the logic from it seems to be the same in `dumbo`, `mmds` and `api_server` too (https://github.com/firecracker-microvm/firecracker/blob/master/src/api_server/src/parsed_request.rs#L94). Since building a response based on some specific inputs is rather application specific, it would make the most sense to have this in a helper crate from Firecracker, i.e. the `utils` one. But this way, `utils` will have an extra dependency on `micro-http` and we don't think it's worth it. It's fine to leave `build_response()` duplicated in `mmds` and `dumbo`.

[the remark this answers is not in the document]
Comment (moricho) [reply src/dumbo/src/tcp/endpoint.rs:307]: I fixed!
